src/learner.cc

```diff
@@ -193,60 +190,129 @@ void GenericParameter::ConfigureGpuId(bool require_gpu) {
 #endif  // defined(XGBOOST_USE_CUDA)
 }
 
+
 using XGBAPIThreadLocalStore =
     dmlc::ThreadLocalStore<std::map<Learner const *, XGBAPIThreadLocalEntry>>;
 
-/*!
- * \brief learner that performs gradient boosting for a specific objective
- * function. It does training and prediction.
- */
-class LearnerImpl : public Learner {
- public:
-  explicit LearnerImpl(std::vector<std::shared_ptr<DMatrix> > cache)
-      : need_configuration_{true} {
-    monitor_.Init("Learner");
-    for (std::shared_ptr<DMatrix> const& d : cache) {
-      cache_.Cache(d, GenericParameter::kCpuId);
+struct LearnerAttributes {
+  /*! \brief Constant string identifying a passed in parameter is a metric name. */
+  static std::string const kEvalMetric;  // NOLINT
+  /*! \brief A global storage for internal prediction. */
+  PredictionContainer cache;
```

Comment:
If any of these variables are only used by one of the learner classes and do not need to be shared, move them to be members of that class. The idea is that variables are not shared if they do not need to be.